### PR TITLE
Discover local running pods from API server and docker daemon

### PR DESCRIPTION
**Pull request by liwenwu-amazon**
*Issue #, if available:*

*Description of changes:*
Today, during ipamD initialization,  it discover all current running Pods from kubelet's insecure port 10255. This PR change it to discover current running pods from K8S API server and docker daemon.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: misc/aws-k8s-cni.yaml [hunk 1]
```diff
@@ -18,18 +55,23 @@ spec:
...
+      serviceAccountName: aws-node
       hostNetwork: true
       tolerations:
       - key: node-role.kubernetes.io/master
         effect: NoSchedule
       - key: CriticalAddonsOnly
         operator: Exists
       containers:
-      - image: 602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:0.1.4
+      - image: 602401143452.dkr.ecr.us-west-2.amazonaws.com/amazon-k8s-cni:1.1.1-test
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply misc/aws-k8s-cni.yaml]: good point on having misc/0.1/ and misc/0.2

File: pkg/k8sapi/discovery.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	controller *controller
+	kubeClient kubernetes.Interface
+	myNodeName string
+	synced     bool
+}
+
+// NewController creats a new DiscoveryController
+func NewController(clientset kubernetes.Interface) *Controller {
+	return &Controller{kubeClient: clientset,
+		myNodeName: os.Getenv("MY_NODE_NAME"),
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply pkg/k8sapi/discovery.go]: this is from downward API.

File: pkg/docker/docker.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// ContainerInfo provides container information
+type ContainerInfo struct {
+	ID     string
+	Name   string
+	K8SUID string
+}
+
+func GetRunningContainers() ([]ContainerInfo, error) {
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply pkg/docker/docker.go]: not today especially when ipamdD can not query kubelet

File: ipamd/ipamd.go [hunk 4]
```diff
@@ -215,6 +220,48 @@ func (c *IPAMContext) nodeInit() error {
...
+	}
+
+	containers, err := docker.GetRunningContainers()
+
+	if err != nil {
+		log.Errorf("Not able to get container info from docker: %v", err)
+		return nil, errors.Wrapf(err, "not able to get container info from docker")
+	}
+
+	for i, pod := range pods {
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply ipamd/ipamd.go]: These are only pods running locally on the node. This happens once only after ipamd restart.

File: pkg/k8sapi/discovery.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+	if !d.kubeStateMetricPodDiscovered {
+		return "", "", errors.New("kube-state-metrics pod not discovered")
+	}
+
+	return d.kubeStateMetricPodName, d.kubeStateMetricPodNode, nil
+}
+
+// GetCNIPods return the list of CNI pod names
+func (d *Controller) GetCNIPods() []string {
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply pkg/k8sapi/discovery.go]: these are used in cni-metrics-helper

File: pkg/k8sapi/discovery.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
+// Package k8sapi
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply pkg/k8sapi/discovery.go]: for cni-metrics-helper, I can remove them for now

File: ipamd/ipamd.go [hunk 7]
```diff
@@ -190,7 +193,7 @@ func (c *IPAMContext) nodeInit() error {
 		}
 	}
 
-	usedIPs, err := c.k8sClient.K8SGetLocalPodIPs(c.awsClient.GetLocalIPv4())
+	usedIPs, err := c.getLocalPodsWithRetry()
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply ipamd/ipamd.go]: the return has not changed yet. K8SGetLocalPodIPs(localIP string) ([]*K8SPodInfo, error) before

File: ipamd/ipamd.go [hunk 8]
```diff
@@ -215,6 +220,48 @@ func (c *IPAMContext) nodeInit() error {
 	return nil
 }
 
+func (c *IPAMContext) getLocalPodsWithRetry() ([]k8sapi.K8SPodInfo, error) {
+	var pods []k8sapi.K8SPodInfo
+	var err error
+	retry := 0
+	for {
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply ipamd/ipamd.go]: done

File: ipamd/ipamd.go [hunk 9]
```diff
@@ -215,6 +220,48 @@ func (c *IPAMContext) nodeInit() error {
...
+	if err != nil {
+		log.Errorf("Not able to get container info from docker: %v", err)
+		return nil, errors.Wrapf(err, "not able to get container info from docker")
+	}
+
+	for i, pod := range pods {
+		// needs to find the container ID
+		for _, container := range containers {
+			//e.g. /k8s_POD_worker-hello-5974f49799-q9vct_default_c31721a2-5dfb-11e8-b09c-022ad646a21e_0
+			k8sName := "/k8s_POD_" + pod.Name + "_" + pod.Namespace + "_" + pod.UID + "_0"
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply ipamd/ipamd.go]: I could not find any.  Right now, I have put it the comments above.

File: ipamd/ipamd.go [hunk 10]
```diff
@@ -215,6 +220,48 @@ func (c *IPAMContext) nodeInit() error {
...
+			log.Debugf("Try to find container with name: %v", k8sName)
+			if strings.Compare(container.Name, k8sName) == 0 {
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply ipamd/ipamd.go]: done

File: main.go [hunk 11]
```diff
@@ -20,19 +20,31 @@ import (
...
 	"github.com/aws/amazon-vpc-cni-k8s/ipamd"
 	log "github.com/cihub/seelog"
+
+	"github.com/aws/amazon-vpc-cni-k8s/pkg/k8sapi"
 )
 
 const (
 	defaultLogFilePath = "/host/var/log/aws-routed-eni/ipamd.log"
-	version            = "0.1.4"
+	version            = "1.1.1-test"
```
[the remark this answers is not in the document]
Comment (liwenwu-amazon) [reply main.go]: have changed it to 0.2.1

**Comment by mattlandis** [comment]
Ok.  Looks good. Go ahead and rebase so you can merge.

**Pull request merged**